File: code/matlab/hsmcel_e_4_6.m
function visual(mode, azimuth, elevation, point_A, point_B, point_C, point_D, point_A1, point_B1, point_C1, point_D1, point_E, point_P, point_Q, point_R)
    close all;
    fig = figure('Visible', 'off');
    
    h = sqrt(6);
    
    A = [0, 0, 0];
    D = [-4, 0, 0];
    B = [0, 4, 0];
    C = [-4, 4, 0];
    
    A1 = [-1, 1, h];
    B1 = [-1, 3, h];
    C1 = [-3, 3, h];
    D1 = [-3, 1, h];
    
    E = (B + C) / 2;
    
    t = 0.6; 
    P = t * D1;
    Q = t * B1;
    R = E + t * (C1 - E);
    
    hold on;
    
    plot3([A(1), B(1)], [A(2), B(2)], [A(3), B(3)], 'k--', 'LineWidth', 2);
    plot3([A(1), D(1)], [A(2), D(2)], [A(3), D(3)], 'k--', 'LineWidth', 2);
    plot3([A(1), A1(1)], [A(2), A1(2)], [A(3), A1(3)], 'k--', 'LineWidth', 2);

    plot3([B(1), C(1)], [B(2), C(2)], [B(3), C(3)], 'k-', 'LineWidth', 2);
    plot3([C(1), D(1)], [C(2), D(2)], [C(3), D(3)], 'k-', 'LineWidth', 2);
    
    plot3([A1(1), B1(1)], [A1(2), B1(2)], [A1(3), B1(3)], 'k-', 'LineWidth', 2);
    plot3([B1(1), C1(1)], [B1(2), C1(2)], [B1(3), C1(3)], 'k-', 'LineWidth', 2);
    plot3([C1(1), D1(1)], [C1(2), D1(2)], [C1(3), D1(3)], 'k-', 'LineWidth', 2);
    plot3([D1(1), A1(1)], [D1(2), A1(2)], [D1(3), A1(3)], 'k-', 'LineWidth', 2);
    
    plot3([B(1), B1(1)], [B(2), B1(2)], [B(3), B1(3)], 'k-', 'LineWidth', 2);
    plot3([C(1), C1(1)], [C(2), C1(2)], [C(3), C1(3)], 'k-', 'LineWidth', 2);
    plot3([D(1), D1(1)], [D(2), D1(2)], [D(3), D1(3)], 'k-', 'LineWidth', 2);

    plot3([A(1), D1(1)], [A(2), D1(2)], [A(3), D1(3)], 'k--', 'LineWidth', 1.5);
    plot3([A(1), B1(1)], [A(2), B1(2)], [A(3), B1(3)], 'k--', 'LineWidth', 1.5);
    plot3([C1(1), E(1)], [C1(2), E(2)], [C1(3), E(3)], 'k--', 'LineWidth', 1.5);
    
    
    all_points_labels = {point_A, point_B, point_C, point_D, point_A1, point_B1, point_C1, point_D1, point_E, point_P, point_Q, point_R};
    all_points_coords = {A, B, C, D, A1, B1, C1, D1, E, P, Q, R};
    
    for i = 1:length(all_points_coords)
        pt = all_points_coords{i};
        scatter3(pt(1), pt(2), pt(3), 40, 'k', 'filled');
        text(pt(1)+0.2, pt(2)+0.2, pt(3)+0.2, all_points_labels{i}, 'FontSize', 12, 'FontWeight', 'bold');
    end

    axis equal;
    axis off;
    view(azimuth, elevation);
    
    set(gca, 'Color', 'white');
    set(gcf, 'Color', 'white');
    set(gcf, 'ToolBar', 'none');
    set(gcf, 'MenuBar', 'none');
    
    
    set(gcf, 'Position', [100, 100, 1024, 1024]);

    
    
    if mode == 0
        img_dir = fullfile('..', '..', 'data', 'images');
        if ~exist(img_dir, 'dir')
            mkdir(img_dir);
        end
        img_path = fullfile(img_dir, [mfilename, '.png']);
        frame = getframe(gcf);

        imwrite(frame.cdata, img_path);
        fprintf('Image saved as: %s \n', img_path);
    elseif mode == 1
        vid_dir = fullfile('..', '..', 'data', 'videos');
        if ~exist(vid_dir, 'dir')
            mkdir(vid_dir);
        end
        vid_path = fullfile(vid_dir, [mfilename, '.mp4']);
        video = VideoWriter(vid_path, 'MPEG-4');
        video.FrameRate = 24;
        open(video);

        set(gca, 'CameraViewAngleMode', 'manual');
        set(gca, 'CameraPositionMode', 'manual');
        set(gca, 'CameraTargetMode', 'manual');

        for angle = (azimuth+3):3:(azimuth+360)
            view(angle, elevation);
            frame = getframe(gcf);
            writeVideo(video, frame);
        end

        close(video);
        fprintf('Video saved as: %s \n', vid_path);
    end
    hold off;
    close(fig);
end
    

File: code/python/hsmcel_e_4_6.py
import json
import os
# Set random seed
import random
import argparse
import string
import math

parser = argparse.ArgumentParser()
parser.add_argument('--seed', type=int, default=0, help='Random seed (default: 0)')
parser.add_argument('--mode', type=int, default=0, choices=[0, 1], help='Mode (default: 0)')
args, unknown = parser.parse_known_args()
random.seed(args.seed)

# Scaling factor
len_scaling_factor = round(random.uniform(0.1, 100.0), 1)

# Generate random point names
point_A, point_B, point_C, point_D, point_A1, point_B1, point_C1, point_D1, point_E, point_P, point_Q, point_R = random.sample(string.ascii_uppercase, 12)

# Add result calculation code
import math

def calculate(len_a, len_b, len_h):
    """
    计算正四棱台中特定三角形的面积

    参数:
    len_a (float): 上底边长
    len_b (float): 下底边长
    len_h (float): 棱台的高

    返回:
    float: 三角形的面积
    """
    # 计算分子部分: - (len_b - len_a)**2 * t**2 + (len_b**2 - len_a**2) / 2 * t
    # 由于t = 1时，得到最大面积公式: (len_b + len_a)**2 / 16
    numerator = (len_b + len_a)**2

    # 计算分母部分: 16
    denominator = 16

    # 返回三角形的面积
    return numerator / denominator

# 示例值
len_a = 2.0  # 上底边长
len_b = 4.0  # 下底边长
len_h = math.sqrt(6)  # 棱台的高
#
# # 调用函数计算面积
# result = calculate(len_a, len_b, len_h)
#
# # 输出结果
# print(f"三角形的面积为: {result:.6f}")
# Generate random lengths
len_a = round(len_scaling_factor * float(len_a), 2)
len_b = round(len_scaling_factor * float(len_b), 2)
len_h = round(len_scaling_factor * float(len_h), 2)

# Calculate the result
result = calculate(len_a, len_b, len_h)

# ─── 1. save JSON ─────────────────────────────── 
json_data = {
    "id": "hsmcel_e_4_6",
    "type": 7,
    "level": 3,
    "cn_problem": f"在正四棱台{point_A}{point_B}{point_C}{point_D}-{point_A1}{point_B1}{point_C1}{point_D1}中，上底边长{point_A1}{point_B1} = {len_a}，下底边长{point_A}{point_B} = {len_b}（{len_b} > {len_a} > 0），高为{len_h}，{point_E}为下底{point_B}{point_C}的中点。作平行于底面的截面，与线段{point_A}{point_D1}、{point_A}{point_B1}、{point_C1}{point_E}分别交于{point_P}、{point_Q}、{point_R}，求△{point_P}{point_Q}{point_R}面积的最大值。",
    "en_problem": f"In a regular quadrangular frustum {point_A}{point_B}{point_C}{point_D}-{point_A1}{point_B1}{point_C1}{point_D1}, the upper base edge length {point_A1}{point_B1} = {len_a}, the lower base edge length {point_A}{point_B} = {len_b} ({len_b} > {len_a} > 0), the height is {len_h}, and {point_E} is the midpoint of the lower base {point_B}{point_C}. A cross-section parallel to the base is made, intersecting line segments {point_A}{point_D1}, {point_A}{point_B1}, and {point_C1}{point_E} at points {point_P}, {point_Q}, and {point_R} respectively. Find the max value of the area of triangle {point_P}{point_Q}{point_R}.",
    "solution": f"{result}",
    "image": f"images/{os.path.splitext(os.path.basename(__file__))[0]}.png"
}

# video mode
if args.mode == 1:
    json_data["image"] = f"videos/{os.path.splitext(os.path.basename(__file__))[0]}.mp4"



# Save to jsonl
jsonl_path = os.path.join(os.path.dirname(__file__), "../../data/problem.jsonl")
os.makedirs(os.path.dirname(jsonl_path), exist_ok=True)
with open(jsonl_path, "a", encoding="utf-8") as f:
    f.write(json.dumps(json_data, ensure_ascii=False) + "\n")
    
# ─── 2. save MATLAB command JSONL ─────────────────────────────── 
mode = args.mode
azimuth = (-150 + random.randint(0, 360)) % 360
elevation = (25 + random.randint(0, 360)) % 360

matlab_cmd_jsonl_path = os.path.join(os.path.dirname(__file__), "../../data/matlab_cmd.jsonl")
os.makedirs(os.path.dirname(matlab_cmd_jsonl_path), exist_ok=True)
with open(matlab_cmd_jsonl_path, "a", encoding="utf-8") as f:
    f.write(json.dumps({json_data["id"]: f"hsmcel_e_4_6({mode}, {azimuth}, {elevation}, '{point_A}', '{point_B}', '{point_C}', '{point_D}', '{point_A1}', '{point_B1}', '{point_C1}', '{point_D1}', '{point_E}', '{point_P}', '{point_Q}', '{point_R}')"}, ensure_ascii=False) + "\n")
